packages/preview/src/browser/preview-widget.ts

```diff
@@ -0,0 +1,258 @@
+        this.title.iconClass = DEFAULT_ICON;
+        this.title.closable = true;
+        this.addClass(PREVIEW_WIDGET_CLASS);
+        this.node.tabIndex = 0;
+        this.startScrollSync();
+        this.startDoubleClickListener();
+        this.update();
+        this.toDispose.push(ThemeService.get().onThemeChange(() => this.update()));
+    }
+
+    protected preventScrollNotification: boolean = false;
+    protected startScrollSync(): void {
+        this.node.addEventListener('scroll', throttle(50, (event: UIEvent) => {
+            if (this.preventScrollNotification) {
+                return;
+            }
+            const scrollTop = this.node.scrollTop;
+            this.didScroll(scrollTop);
+        }));
+    }
+
+    protected startDoubleClickListener(): void {
+        this.node.addEventListener('dblclick', (event: MouseEvent) => {
+            if (!(event.target instanceof HTMLElement)) {
+                return;
+            }
+            const target = event.target as HTMLElement;
+            let node: HTMLElement | null = target;
+            while (node && node instanceof HTMLElement) {
+                if (node.tagName === 'A') {
+                    return;
+                }
+                node = node.parentElement;
+            }
+            const offsetParent = target.offsetParent as HTMLElement;
+            const offset = offsetParent.classList.contains(PREVIEW_WIDGET_CLASS) ? target.offsetTop : offsetParent.offsetTop;
+            this.didDoubleClick(offset);
+        });
+    }
+
+    onActivateRequest(msg: Message): void {
+        super.onActivateRequest(msg);
+        if (this.node.children.length > 0) {
+            (this.node.children.item(0) as HTMLElement).focus();
+        } else {
+            this.node.focus();
+        }
+        this.update();
+    }
+
+    onUpdateRequest(msg: Message): void {
+        super.onUpdateRequest(msg);
+        this.performUpdate();
+    }
+
+    protected async performUpdate(): Promise<void> {
+        if (!this.resource) {
+            return;
+        }
+        const uri = this.resource.uri;
+        const document = this.workspace.textDocuments.find(d => d.uri === uri.toString());
+        const content: MaybePromise<string> = document ? document.getText() : this.resource.readContents();
+        const contentElement = await this.render(await content, uri);
+        this.node.innerHTML = '';
+        if (contentElement) {
+            this.node.appendChild(contentElement);
+            if (this.firstUpdate) {
+                this.firstUpdate();
+                this.firstUpdate = undefined;
+            }
+        }
+    }
+
+    protected async render(content: string, originUri: URI): Promise<HTMLElement | undefined> {
+        if (!this.previewHandler || !this.resource) {
+            return undefined;
+        }
+        return this.previewHandler.renderContent({ content, originUri });
+    }
+
+    storeState(): object {
```

Comment:
The uri is the key for the widget and should be set by the factory. So there is no need for storing it here again.
The store/restore is really just to store state that is heavy to compute (e.g. error markers, git changes, etc.) In the case of preview, it would be the result from the PreviewHandler. But my feeling is that it would be quite heavy to store it in the browser storage and it is not so hard to recompute, so we better just don't implement `StatefulWidget`.

Reply:
Thanks for pointing this out. I've cleaned that up. I've did this similar to the `TerminalWidget` when the preview widget meant to be a singleton. 
Actually, I was not aware that, the `createWidget`'s options are persisted, that's cool!